File: top-interview/array/best-time-to-buy/python/best_time_test.py
import unittest

from best_time import Solution

class TestBestTime(unittest.TestCase):
    def setUp(self):
        self.testCases = [
            {
                'prices': [7,1,5,3,6,4],
                'want': 7,
            },
            {
                'prices': [1,2,3,4,5],
                'want': 4,
            },
            {
                'prices': [7,6,4,3,1],
                'want': 0,
            },
        ]

    def test(self):
        s = Solution()
        for testCase in self.testCases:
            with self.subTest(testCase=testCase):
                self.assertEqual(
                    s.maxProfit(testCase['prices']),
                    testCase['want']
                )

if __name__ == '__main__':
    unittest.main()

File: top-interview/array/best-time-to-buy/go/best_time_test.go
package besttime

import "testing"

func TestMaxProfit(t *testing.T) {
	var testCases = []struct {
		prices []int
		want   int
	}{
		{
			[]int{7, 1, 5, 3, 6, 4},
			7,
		},
		{
			[]int{1, 2, 3, 4, 5},
			4,
		},
		{
			[]int{7, 6, 4, 3, 1},
			0,
		},
	}

	for idx, testCase := range testCases {
		if r := maxProfit(testCase.prices); r != testCase.want {
			t.Errorf("%d : %v, %d != %d", idx, testCase, r, testCase.want)
		}
	}
}
